pkg/jobs/jobspb/jobs.proto

```diff
@@ -214,7 +214,12 @@ message BackupDetails {
     (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.ID"
   ];
 
-  // NEXT ID: 20;
+  // HasLockedLocation is set to true once the backup job has checked that are no concurrent backups
+  // writing to the same location, and has written a lock file to prevent any other backups from
+  // writing to the location.
+  bool has_locked_location = 20;
```

Comment:
I wonder if we need this.  Namely, the source of truth is whether the backup location has been locked is whether the lockfile exists.  We could always check for the lockfile and only use the fallback checks if we don't find an existing lockfile for our job.

Reading through the code I believe it is fine as-is, and this does save us from a little extra work. But keeping state in sync between the job details and the bucket is partially the original cause of this bug so it might be worth thinking about whether we need to do that here.

Reply:
Replying to myself, I think the main thing this helps with is the mixed-version case.

Reply:
As discussed offline, in a mixed version state saves us because we still read`BACKUP-CHECKPOINT` to block concurrent backups, and so a job resumption could result in a job reading its own checkpoint file. To protect against this we gate the step behind this bool.

Reply:
@stevendanna actually I've convinced myself that with the new logic where we use the presence of our lock file as a signal of whether or not to repeat the check+write I don't think we need the bool. To deal with the mixed version case, we can simply rely on `details.URI` being non-empty since it would have been populated during planning.